File: skills/faion/knowledge/pro/dev/software-developer/ruby-activerecord/templates/order.rb
# purpose: ActiveRecord model with enum, scopes, includes-friendly associations
# consumes: see content/02-output-contract.xml inputs
# produces: artefact conforming to content/02-output-contract.xml
# depends-on: content/01-core-rules.xml
# token-budget-impact: ~400 tokens when loaded as context

class Order < ApplicationRecord
  belongs_to :customer
  has_many :items, class_name: "OrderItem", dependent: :destroy

  enum status: { pending: 0, paid: 1, shipped: 2, cancelled: 3 }, _prefix: true

  validates :status, presence: true
  validates :total_cents, numericality: { greater_than_or_equal_to: 0 }

  scope :recent, -> { where("created_at >= ?", 30.days.ago) }
  scope :for_user, ->(user) { where(customer_id: user.customer_id) }
  scope :priced, -> { where("total_cents > 0") }

  def total
    Money.new(total_cents, currency)
  end

  def cancel!
    transaction do
      update!(status: :cancelled, cancelled_at: Time.current)
      items.each(&:release_inventory!)
    end
  end
end

File: .archive/knowledge/dev/php-eloquent/templates/Order.php
// purpose: Eloquent model with $fillable, $casts, scopes, and typed relations
// consumes: see content/02-output-contract.xml inputs
// produces: artefact conforming to content/02-output-contract.xml
// depends-on: content/01-core-rules.xml
// token-budget-impact: ~450 tokens when loaded as context

<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Order extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'customer_id',
        'status',
        'note',
        'total_cents',
    ];

    protected $casts = [
        'total_cents' => 'integer',
        'paid_at' => 'datetime',
        'metadata' => 'array',
    ];

    public function customer(): BelongsTo
    {
        return $this->belongsTo(Customer::class);
    }

    public function items(): HasMany
    {
        return $this->hasMany(OrderItem::class);
    }

    public function scopeRecent(Builder $query): Builder
    {
        return $query->where('created_at', '>=', now()->subDays(30));
    }

    public function scopePaid(Builder $query): Builder
    {
        return $query->whereNotNull('paid_at');
    }
}
